### DISCUSSION TITLE
Option to disable config live-reloading

### DISCUSSION THREAD

**Discussion by HardGraphite** [YaLTeR/niri#219] [Ideas]
The config live-reloading feature is nice for debugging and exploring. However, I'm afraid it is not very useful for daily use, especially considering that it is implemented by polling the file metadata every 500 ms.

Is it a good idea to add an option allowing user to disable this feature? Maybe it can be called `live_config_reload`, just like what is done in Alacritty.

**Comment by YaLTeR** [comment]
I mean I guess I could add a flag to disable it, I'm just not sure when 500 ms metadata polling could be this big of a problem.

**Reply by HardGraphite** [reply]
Haha

I must admit that I've never tested the cost of the polling. I'm trying niri on an old laptop so I want to minimize the unnecessary costs. Anyway, it's not a crucial task.

By the way, I really like the way niri works. I didn't know PaperWM before, and this is the first time I learn about scrollable-tiling WM.

**Comment by kchibisov** [comment]
The thing about cost here is that niri always runs and refreshes/polls one way or another, where e.g. alacritty could be obscured thus suspended and never receive events. So it's the same as why kernel polls input devices, etc.

**Comment by pilot0904** [comment]
I know this is an old post. Just started using Niri and really liked it. Running Niri in Nixos, came across an issue where the config.kdl was locked somehow and I have to delete the file so Nixos home manager can complete the build by copying over the new config. I suspect that it is the file polling. I deleted the file and the home manager built without an issue and copy over the new file. Plus I am running this on an old laptop, so any help to reduce polling and battery usage, no matter how small would help a lot. It would be nice to have an option in config or a key bind to enable/disable config file auto-reload and the ability to force a reload.

**Reply by YaLTeR** [reply]
Polling doesn't keep the file open so it's not the problem here

**Comment by voidstar-null** [comment]
I would also like this to be implemented. In the Hyprland docs, its stated that disabling live reloading config might help saving some battery.
![image](https://github.com/user-attachments/assets/f25c26a9-0c28-422a-8b60-940214a54c01)

If an option can help squeeze some extra minutes out of the laptop battery, I think it would make sense to implement this, if that's easy to implement. Apart from this, the error pop is quite annoying too when the editor has autosave on text change enabled.

**Reply by YaLTeR** [reply]
Drawing conclusions based on docs of a different compositor is certainly an interesting strategy

**Reply by voidstar-null** [reply]
I'm so sorry, but doesn't live reloading work the same in all compositors. I mean don't all compositors poll after a specific duration to check the config and reload it ?? Please right me if I'm wrong .

Also there's this too T_T .
> Apart from this, the error pop is quite annoying too when the editor has autosave on text change enabled.

**Reply by voidstar-null** [reply]
Imo this feature should be prioritized as disabling live reloading might help save on cpu cycles which would do a great deal in saving those 0.00001 seconds of battery life lost by polling.

/s

**Reply by YaLTeR** [reply]
> I'm so sorry, but doesn't live reloading work the same in all compositors. I mean don't all compositors poll after a specific duration to check the config and reload it ?? Please right me if I'm wrong .

For one, it can be poll or something inotify-based. Poll can have different interval and different amount of code running on every poll. Then, it's quite hard to estimate battery impact in general because it depends on many factors.

Niri polls twice a second with the bare minimum code on wakeup. So far I've only heard some anecdotal evidence that its battery life is better than sway. I can't draw much in terms of conclusions from that, other than niri definitely doesn't do anything catastrophic with power use.

**Reply by voidstar-null** [reply]
As I mentioned earlier, live reload of config can be annoying too while editing the config file , and the editor has autoSave enabled, For example,  in my case, I have an autocommand for textChanged event for my neovim setup (cuz of some annoying freezes recently), so niri reloads the config after each character insertion and due to this, the error popup appears a million times while editing the config, which is kinda annoying to deal with.

**Reply by YaLTeR** [reply]
Well that's honestly just a self-inflicted problem? I don't think editors are expected to auto-save on every character.

**Reply by voidstar-null** [reply]
Idk but the last time I used VsCode, its autosave worked similarly too, but with a small delay.

**Reply by voidstar-null** [reply]
If this is too much of a work for you, then yeah maybe its not worth implementing.

**Comment by itshog** [comment]
Even if disabling live reloading makes only a small difference in terms of performance and resource usage I believe it could be worth it when running Niri on very old or low-end hardware. Also, as said before, error popups can be very annoying when editing the config file and accidentally saving the file or maybe making changes that break the config when testing new features and options. I believe the best course of action would be (if this isn't too much work) to make disabling live reloading an option for those who want it, keeping the default as of now and adding a `niri msg reload` command similar to `swaymsg reload` in Sway for manually reloading the config file.

**Reply by HigherOrderLogic** [reply]
> Also, as said before, error popups can be very annoying when editing the config file and accidentally saving the file or maybe making changes that break the config when testing new features and options.

A new config field was added to disable the failed to parse config recently.

> adding a `niri msg reload` command similar to `swaymsg reload` in Sway for manually reloading the config file.

It has also been implemented.

**Reply by itshog** [reply]
> > adding a `niri msg reload` command similar to `swaymsg reload` in Sway for manually reloading the config file.
> 
> It has also been implemented.

Sorry, missed that one! Anyway the point presented here about performance/resource usage still stands. Especially if manual reloading is already implemented I think disabling live reloading makes a lot of sense when paired with that feature (Sorry for my ignorance, but I don't believe manual reloading is that useful when the config auto-reloads anyway...whereas if live reloading is disabled a command for manually reloading the config is really nice to have)

**Comment by Treeniks** [comment]
Maybe to add another non-performance related usecase: I have configurations for multiple different setups in a single dotfiles repo and it's not uncommon for me to to have machine-specific uncommitted lines (e.g. a `desktop.kdl` and a `laptop.kdl` config, as well as a global `config.kdl` with defaults and in which I just include the respective machine's config). Sometimes, when pulling changes I made on another machine, I have to temporarily stash some of these includes so git doesn't get conflicts, which can cause my entire configuration to get unloaded.

Now I could change my setup to be less scuffed, but more generally: Accidentally commenting out an important include can be much more devastating than getting an error message, as that include might've defined `output` properties.
